sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -714,6 +764,197 @@ cdef class Splitter:
+            Y_DTYPE_C sum_gradients_bin
+            Y_DTYPE_C sum_hessians_bin
+            Y_DTYPE_C loss_current_node
+            # Reduces the effect of noises in categorical features,
+            # especially for categoires with few data
+            # TODO: Make this user adjustable?
+            unsigned int CAT_SMOOTH = 10
+            # Used for find best split
+            unsigned int MAX_CAT_THRESHOLD = 32
+            unsigned int max_num_cat
+            # holds directional information
+            int[2] find_direction, start_position
+            int direction, position
+            Y_DTYPE_C sum_gradient_left, sum_hessian_left
+            Y_DTYPE_C sum_gradient_right, sum_hessian_right
+            unsigned int n_samples_left, n_samples_right
+            unsigned int i, j
+            Y_DTYPE_C gain
+            Y_DTYPE_C best_gain = -1.0
+            int best_direction
+            unsigned char found_better_split = False
+            Y_DTYPE_C best_sum_hessian_left
+            Y_DTYPE_C best_sum_gradient_left
+            unsigned int best_n_samples_left
+            unsigned int best_sort_thres
+
+        cat_sort_infos = <categorical_info *> malloc(
+            (end + has_missing_values) * sizeof(categorical_info))
+
+        # filter out categoires based on CAT_SMOOTH
+        for bin_idx in range(end):
+            if feature_hist[bin_idx].count >= CAT_SMOOTH:
+                cat_sort_infos[used_bin].bin_idx = bin_idx
+                sum_gradients_bin = feature_hist[bin_idx].sum_gradients
+                if self.hessians_are_constant:
+                    sum_hessians_bin = feature_hist[bin_idx].count
+                else:
+                    sum_hessians_bin = feature_hist[bin_idx].sum_hessians
+
+                cat_sort_infos[used_bin].value = \
+                    sum_gradients_bin / (sum_hessians_bin + CAT_SMOOTH)
+                used_bin += 1
+
+        # check missing bin
+        if has_missing_values:
+            if feature_hist[missing_values_bin_idx].count >= CAT_SMOOTH:
+                cat_sort_infos[used_bin].bin_idx = missing_values_bin_idx
+                sum_gradients_bin = \
+                    feature_hist[missing_values_bin_idx].sum_gradients
+                if self.hessians_are_constant:
+                    sum_hessians_bin = \
+                        feature_hist[missing_values_bin_idx].count
+                else:
+                    sum_hessians_bin = \
+                        feature_hist[missing_values_bin_idx].sum_hessians
+
+                cat_sort_infos[used_bin].value = \
+                    sum_gradients_bin / (sum_hessians_bin + CAT_SMOOTH)
+                used_bin += 1
+
+        # not enough categories to form a split
+        if used_bin <= 1:
+            free(cat_sort_infos)
+            return
+
+        qsort(cat_sort_infos, used_bin, sizeof(categorical_info),
+              compare_cat_infos)
+
+        max_num_cat = min(MAX_CAT_THRESHOLD, (used_bin + 1) / 2)
```

Comment:
hmm, two questions:

- why do we want to limit to `(used_bin + 1) / 2`? it seems to me that the rest of the splits, i.e. the other half, are still relevant?
- why do we want to limit to MAX_CAT_THRESHOLD? According to https://github.com/microsoft/LightGBM/issues/699#issuecomment-319565499, this is used to reduce communication cost in parallel learning, but I don't understand that

Reply:
Ok so I guess for the first point, we're still considering the rest of the splits because we consider both directions, each ending at the middle.

But then the question becomes: why do we need to do that? It seems that the reason is to limit the loop below `for i in range(best_sort_thres + 1)`, but I can't say I'm really convinced?

Reply:
Assuming we have `MAX_CAT_THRESHOLD=32` and it does not reach the center, going in both directions allows us to consider 32 categories from the left, and then 32 categories from the right.

If we get rid of `MAX_CAT_THRESHOLD`, then I do not see a point to going in both directions.